proto/substrait/algebra.proto

```diff
@@ -178,6 +178,39 @@ message JoinRel {
   substrait.extensions.AdvancedExtension advanced_extension = 10;
 }
 
+// A time-series variant of the multi JOIN relational operator left-join-right(s), which joins on an ordered key using inexact matching
+message AsOfJoinRel {
+  RelCommon common = 1;
+  Rel left = 2;
+  repeated Rel right = 3;
+  Expression expression = 4;
+  Expression post_join_filter = 5;
+
+  JoinType type = 6;
+
+  enum JoinType {
+    JOIN_TYPE_UNSPECIFIED = 0;
+    JOIN_TYPE_INNER = 1;
+    JOIN_TYPE_OUTER = 2;
+    JOIN_TYPE_LEFT = 3;
+    JOIN_TYPE_RIGHT = 4;
+    JOIN_TYPE_SEMI = 5;
+    JOIN_TYPE_ANTI = 6;
+    // This join is useful for nested sub-queries where we need exactly one tuple in output (or throw exception)
+    // See Section 3.2 of https://15721.courses.cs.cmu.edu/spring2018/papers/16-optimizer2/hyperjoins-btw2017.pdf
+    JOIN_TYPE_SINGLE = 7;
+  }
+
+  Expression.Literal tolerance = 7;
+
+  oneof on_type {
+    Expression.FieldReference on = 8;
+    // Reserve next tags for future on_type alternatives
+  }
+
+  substrait.extensions.AdvancedExtension advanced_extension = 100;
```

Comment:
What is this for?

Reply:
See [docs](https://substrait.io/extensions/#advanced-extensions). It allows producer/consumer pairs that know about each other to pass extra information along with a relation, distinguished by being either optimization-only (may be ignored by the consumer) or "enhancements" that change the semantics of the relation somehow (may not be ignored).

That being said, I don't know why some relations have a dedicated field and others don't. All relations have access to an advanced extension field (multiple, actually) via `RelCommon`.